loading_cache: force minimum size of unprivileged

avelanarius opened the pull request:
This series enforces a minimum size of the unprivileged section when
performing `shrink()` operation.

When the cache is shrunk, we still drop entries first from unprivileged
section (as before this commit), however, if this section is already small
(smaller than `max_size / 2`), we will drop entries from the privileged
section.

This is necessary, as before this change the unprivileged section could
be starved. For example if the cache could store at most 50 entries and 
there are 49 entries in privileged section, after adding 5 entries (that would 
go to unprivileged section) 4 of them would get evicted and only the 5th one 
would stay. This caused problems with BATCH statements where all 
prepared statements in the batch have to stay in cache at the same time 
for the batch to correctly execute.

To correctly check if the unprivileged section might get too small after
dropping an entry, `_current_size` variable, which tracked the overall size
of cache, is changed to two variables: `_unprivileged_section_size` and 
`_privileged_section_size`, tracking section sizes separately.

New tests are added to check this new behavior and bookkeeping of the section
sizes. A test is added, that sets up a CQL environment with a very small 
prepared statement cache, reproduces issue in #10440 and stresses the cache.

Fixes #10440.

avelanarius commented:
Sorry, marking as draft as there is a problem in this condition:
`_unprivileged_lru_list.size() > minimum_unprivileged_size`.

The value on the right is _max_size / 2, but in some instances it refers to memory size, not number of entries (lhs is number of entries).

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/200/

avelanarius commented:
v2:
- The incorrect condition (`_unprivileged_lru_list.size() > minimum_unprivileged_size`) which compared different units: `number of entries [count] > cache size [memory footprint]` is fixed. Now it compares the section size vs cache size (both "memory footprint").
- To perform this comparison, the size of each section (privileged/unprivileged) is now tracked separately.
- Added a new test that inserts entries to the cache with different sizes and checks if the size of sections are correct after each operation.
- Added a new test, that sets up a CQL environment with a very small prepared statement cache, reproduces issue in #10440 and stresses the cache.

avelanarius commented on test/boost/loading_cache_test.cc:
```diff
@@ -428,6 +429,163 @@ SEASTAR_TEST_CASE(test_loading_cache_eviction_unprivileged) {
+            loading_cache.get_ptr(i, load_len10).discard_result().get();
+        }
+        BOOST_REQUIRE_EQUAL(loading_cache.privileged_section_memory_footprint(), 6);
+        // We shrink the cache BEFORE adding element,
+        // so after adding the element, the cache
+        // can exceed max_size...
+        BOOST_REQUIRE_EQUAL(loading_cache.unprivileged_section_memory_footprint(), 100); 
+        BOOST_REQUIRE_EQUAL(loading_cache.size(), 12);
```
While writing this test, I saw another "problem" introduced in 1a9c6d9fd3af5b9dd06bc51128fdcd3185456c70. That commit also changed that the `shrink()` is called before adding a new entry, not after adding a new entry. (The reasoning was that if we just added an only element to unprivileged section, shrinking the cache would evict just added entry).

The effect of such change is that the size of the cache can get larger than `_max_size` by the last entry added.

Decided not to pollute this PR by fixing too many issues.

avikivity replied:
Don't think it needs to be changed. It's not like _max_size is a perfect size and _max_size+1 will cause the sun to collapse into a black hole - we need some limit, and it's not too bad if the limit is a little fuzzy.

(it's much more interesting to convert this limit from a count to a byte size, since that's what we really want to limit, but also much harder)

avelanarius replied:
> (it's much more interesting to convert this limit from a count to a byte size, since that's what we really want to limit, but also much harder)

What do you mean by "this limit"? `_max_size` is the "byte size" ("memory footprint") limit, not number of entries.

This brings me to the next question: maybe we should change the naming we use in `loading_cache`: `_max_size` changed to `_max_memory_footprint` (and others similarly)? In my first attempt, I mistook `_max_size` for number of entires and wrote wrong code. This also leads to a strange code like this - current master:

https://github.com/scylladb/scylla/blob/aee94b7176aad5dfb1b302c1a64b257c6dc4fe7a/utils/loading_cache.hh#L377-L384

`size()` returning number of entires, `memory_footprint()` returns the memory footprint by reading from... `_current_size` and `timestamped_val::size()` returns the memory footprint.

avikivity replied:
> > (it's much more interesting to convert this limit from a count to a byte size, since that's what we really want to limit, but also much harder)
> 
> What do you mean by "this limit"? `_max_size` is the "byte size" ("memory footprint") limit, not number of entries.
> 

I meant _max_size, I incorrectly assumed it was an entry count limit.

> This brings me to the next question: maybe we should change the naming we use in `loading_cache`: `_max_size` changed to `_max_memory_footprint` (and others similarly)? In my first attempt, I mistook `_max_size` for number of entires and wrote wrong code.

Agree. _list.size() really smells like an element count.

> This also leads to a strange code like this - current master:
> 
> https://github.com/scylladb/scylla/blob/aee94b7176aad5dfb1b302c1a64b257c6dc4fe7a/utils/loading_cache.hh#L377-L384
> 
> `size()` returning number of entires, `memory_footprint()` returns the memory footprint by reading from... `_current_size` and `timestamped_val::size()` returns the memory footprint.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/205/

avikivity commented:
> v2:
> 
> * The incorrect condition (`_unprivileged_lru_list.size() > minimum_unprivileged_size`) which compared different units: `number of entries [count] > cache size [memory footprint]` is fixed. Now it compares the section size vs cache size (both "memory footprint").


This is subtly problematic. A very large entry in one section can dominate and cause it to only hold one entry. But this is an edge case (and I don't believe we really calculate the memory footprint).
